src/azul/tdr.py

```diff
@@ -195,39 +200,44 @@ def emulate_bundle(self,
             link_type = link['link_type']
             if link_type == 'process_link':
                 entities[link['process_type']].add(link['process_id'])
-                for catgeory in ('input', 'output', 'protocol'):
-                    for entity_ref in link[catgeory + 's']:
-                        entities[entity_ref[catgeory + '_type']].add(entity_ref[catgeory + '_id'])
+                for category in ('input', 'output', 'protocol'):
+                    for entity_ref in link[category + 's']:
+                        entities[entity_ref[category + '_type']].add(entity_ref[category + '_id'])
             elif link_type == 'supplementary_file_link':
                 # For MVP, only project entities can have associated supplementary files.
                 entity = link['entity']
-                if entity['entity_type'] != 'project' or entity['entity_id'] != links_row['project_id']:
-                    raise ValueError(f'Supplementary file not associated with bundle project: {entity}')
+                require(entity['entity_type'] == 'project' and entity['entity_id'] == links_row['project_id'],
+                        f'Supplmentary files may only be associated with the bundle project: {entity}')
                 for supp_file in link['files']:
                     entities['supplementary_file'].add(supp_file['file_id'])
             else:
-                raise ValueError(f'Unexpected link_type: {link_type}')
+                raise RequirementError(f'Unexpected link_type: {link_type}')
 
-        for entity_type, entity_ids in entities.items():
+        def retrieve_rows(entity_type: str, entity_ids: Set[str]):
             pk_column = entity_type + '_id'
             non_pk_columns = bundler.data_columns if entity_type.endswith('_file') else bundler.metadata_columns
             columns = ', '.join(non_pk_columns | {pk_column})
-            # It's ~32% faster to consolidate queries than to fetch entities separately.
-            # Better way to write this in standardSQL would be
-            # `id IN UNNEST([id1, id2...])` but IDK how to write an UNNEST
-            # function in TinyQuery.
             uuid_in_list = ' OR '.join(f'{pk_column} = "{entity_id}"' for entity_id in entity_ids)
             rows = self._query_latest_version(f'''
                 SELECT {columns}
                 FROM {self.target.name}.{entity_type}
                 WHERE {uuid_in_list}
             ''', group_by=pk_column)
             log.info('Retrieved %s %s entities', len(rows), entity_type)
+            return rows
+
+        with ThreadPoolExecutor(max_workers=len(entities)) as executor:
```

Comment:
Use a TP with a configurable maximum size like the DSS repo plugin does.

We should probably use the same env variable for both. It may need to be renamed.